In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df_ = pd.read_csv("/Users/ibrahimyucel/Downloads/GAIN ENERGY Ibrahim Yucel 2 Case Study/veri/EBER_RES_birlesik_2024.csv")
df = df_.copy()

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Tarih                 8784 non-null   object 
 1   Ay                    8784 non-null   int64  
 2   Saat                  8784 non-null   object 
 3   PTF                   8784 non-null   float64
 4   SMF                   8784 non-null   float64
 5   Poz_DF                8784 non-null   float64
 6   Neg_DF                8784 non-null   float64
 7   KGUP                  8784 non-null   float64
 8   URETIM                8784 non-null   float64
 9   Dengesizlik           8784 non-null   float64
 10  GOP_Geliri            8784 non-null   float64
 11  Dengesizlik_Tutarı    8784 non-null   float64
 12  Net_Gelir             8784 non-null   float64
 13  Dengesizlik_Maliyeti  8784 non-null   float64
 14  Birim_DM              8784 non-null   float64
dtypes: float64(12), int64

In [5]:
df.head()

Tarih  Ay   Saat      PTF      SMF     Poz_DF     Neg_DF  KGUP  \
0  2024-01-01   1  00:00  1299.98   560.00   543.2000  1338.9794  3.44   
1  2024-01-01   1  01:00  1299.98  1299.98  1260.9806  1338.9794  5.08   
2  2024-01-01   1  02:00  1248.54  1248.54  1211.0838  1285.9962  6.16   
3  2024-01-01   1  03:00  1299.98  1250.00  1212.5000  1338.9794  7.08   
4  2024-01-01   1  04:00  1200.00  1150.00  1115.5000  1236.0000  8.89   

   URETIM  Dengesizlik  GOP_Geliri  Dengesizlik_Tutarı     Net_Gelir  \
0    14.0        10.56   4471.9312         5736.192000  10208.123200   
1    19.0        13.92   6603.8984        17552.849952  24156.748352   
2    28.0        21.84   7691.0064        26450.070192  34141.076592   
3    32.0        24.92   9203.8584        30215.500000  39419.358400   
4    24.0        15.11  10668.0000        16855.205000  27523.205000   

   Dengesizlik_Maliyeti    Birim_DM  
0           7991.596800  570.828343  
1            542.871648   28.572192  
2            818.043408   29.215836  
3           2180.001600   68.125050  
4           1276.795000   53.199792

In [ ]:
df['Tarih'] = pd.to_datetime(df['Tarih'])
print(f"\nDönüşüm tamamlandı!")

print(f"Yeni veri tipi: {df['Tarih'].dtype}")
print("Örnek değerler:")
df['Tarih'].head()


Dönüşüm tamamlandı!
Yeni veri tipi: datetime64[ns]
Örnek değerler:


0   2024-01-01
1   2024-01-01
2   2024-01-01
3   2024-01-01
4   2024-01-01
Name: Tarih, dtype: datetime64[ns]

In [18]:
print("=== Veri Kalite Kontrolü ===")
print(f"Toplam satır: {len(df)}")
print(f"Eksik değerler: {df.isnull().sum().sum()}")
print(f"Duplikat satırlar: {df.duplicated().sum()}")

# Saat kontrolü (24 saat × 366 gün = 8784)
print(f"\n=== Saat Dağılımı ===")
print(f"Benzersiz saat sayısı: {df['Saat'].nunique()}")
print(f"Benzersiz tarih sayısı: {df['Tarih'].nunique()}")

# Değer aralıkları
print(f"\n=== Değer Aralıkları ===")
print(f"PTF: {df['PTF'].min():.2f} - {df['PTF'].max():.2f}")
print(f"SMF: {df['SMF'].min():.2f} - {df['SMF'].max():.2f}")
print(f"KGÜP: {df['KGUP'].min():.2f} - {df['KGUP'].max():.2f}")
print(f"Üretim: {df['URETIM'].min():.2f} - {df['URETIM'].max():.2f}")

# Dengesizlik maliyeti özeti
print(f"\n=== Dengesizlik Maliyeti ===")
print(f"Toplam DM: {df['Dengesizlik_Maliyeti'].sum():.2f} TL")
print(f"Ortalama Birim DM: {df['Birim_DM'].mean():.4f} TL/MWh")
print(f"Toplam Üretim: {df['URETIM'].sum():.2f} MWh")
print("*" * 20)
print("NET GELİR: ", df["Net_Gelir"].sum())

=== Veri Kalite Kontrolü ===
Toplam satır: 8782
Eksik değerler: 0
Duplikat satırlar: 0

=== Saat Dağılımı ===
Benzersiz saat sayısı: 24
Benzersiz tarih sayısı: 366

=== Değer Aralıkları ===
PTF: 0.00 - 3000.00
SMF: 0.00 - 3000.00
KGÜP: 0.00 - 62.40
Üretim: 0.00 - 67.91

=== Dengesizlik Maliyeti ===
Toplam DM: 20901867.67 TL
Ortalama Birim DM: 240.1820 TL/MWh
Toplam Üretim: 203779.27 MWh
********************
NET GELİR:  423914447.51297605


In [9]:
df[df["URETIM"] < 0]

Tarih  Ay   Saat      PTF      SMF     Poz_DF     Neg_DF  KGUP  \
3897 2024-06-11   6  09:00  2649.99  2691.05  2570.4903  2771.7815   0.6   
3898 2024-06-11   6  10:00  2639.98  2691.05  2560.7806  2771.7815   0.4   

      URETIM  Dengesizlik  GOP_Geliri  Dengesizlik_Tutarı    Net_Gelir  \
3897   -0.55        -1.15    1589.994        -3187.548725 -1597.554725   
3898   -0.45        -0.85    1055.992        -2356.014275 -1300.022275   

      Dengesizlik_Maliyeti    Birim_DM  
3897            140.060225 -254.654955  
3898            112.031275 -248.958389

In [10]:
df = df[df["URETIM"] >= 0]
df.shape

(8782, 15)

In [12]:
# Aylık agregasyon ve temel performans metrikleri
df_monthly = df.groupby('Ay').agg({
    'KGUP': 'sum',
    'URETIM': 'sum', 
    'Dengesizlik': 'sum',
    'GOP_Geliri': 'sum',
    'Dengesizlik_Tutarı': 'sum',
    'Net_Gelir': 'sum',
    'Dengesizlik_Maliyeti': 'sum'
}).round(2)

# Aylık performans metrikleri
df_monthly['Dengesizlik_Orani'] = (df_monthly['Dengesizlik'] / df_monthly['KGUP'] * 100).round(4)
df_monthly['Birim_DM_Aylik'] = (df_monthly['Dengesizlik_Maliyeti'] / df_monthly['URETIM']).round(4)

# Kapasite faktörü hesaplama
days_per_month = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]  # 2024 artık yıl
max_possible_production = [62.4 * 24 * days for days in days_per_month]
df_monthly['Kapasite_Faktoru'] = (df_monthly['URETIM'] / max_possible_production * 100).round(2)

df_monthly['DM_Gelir_Orani'] = (df_monthly['Dengesizlik_Maliyeti'] / df_monthly['Net_Gelir'] * 100).round(4)

print("=== EBER RES - Aylık Performans Özeti ===")
df_monthly

=== EBER RES - Aylık Performans Özeti ===


KGUP    URETIM  Dengesizlik   GOP_Geliri  Dengesizlik_Tutarı  \
Ay                                                                     
1   22452.99  20442.00     -2010.99  41894565.79         -5976717.72   
2   11630.61  13524.00      1893.39  21127661.03          2281835.44   
3   13649.83  13511.59      -138.24  29930948.63         -1765354.79   
4   23644.19  24483.88       839.69  43032312.58         -1970306.31   
5   12700.89  14337.92      1637.03  26964807.19          1925187.31   
6   15078.27  14239.72      -838.55  31297593.55         -3271364.48   
7   21470.40  21560.31        89.91  53129583.20          -713197.91   
8   14315.40  15441.92      1126.52  36250005.08          1484653.17   
9    8596.59   8419.04      -177.55  20874914.29         -1792176.93   
10  19435.37  21023.54      1588.17  44444199.41          1571645.11   
11  17627.86  16547.83     -1080.03  41067907.55         -4043482.11   
12  22719.60  20247.52     -2472.08  53989301.12         -7820072.69   

      Net_Gelir  Dengesizlik_Maliyeti  Dengesizlik_Orani  Birim_DM_Aylik  \
Ay                                                                         
1   35917848.07            2498113.95            -8.9564        122.2050   
2   23409496.47            1561512.14            16.2794        115.4623   
3   28165593.84            1368400.64            -1.0128        101.2761   
4   41062006.27            2762033.92             3.5514        112.8103   
5   28889994.50            1595060.23            12.8891        111.2477   
6   28026229.07            1265538.70            -5.5613         88.8738   
7   52416385.29            1334557.81             0.4188         61.8988   
8   37734658.25            1466935.76             7.8693         94.9970   
9   19082737.37            1350620.09            -2.0654        160.4245   
10  46015844.52            2251826.36             8.1715        107.1098   
11  37024425.44            1714191.69            -6.1268        103.5901   
12  46169228.42            1733076.38           -10.8808         85.5945   

    Kapasite_Faktoru  DM_Gelir_Orani  
Ay                                    
1              44.03          6.9551  
2              31.14          6.6704  
3              29.10          4.8584  
4              54.50          6.7265  
5              30.88          5.5212  
6              31.69          4.5156  
7              46.44          2.5461  
8              33.26          3.8875  
9              18.74          7.0777  
10             45.28          4.8936  
11             36.83          4.6299  
12             43.61          3.7537

EBER RES Yaz ayları (Haziran-Ağustos) hem operasyonel hem de mali açıdan en başarılı dönem olmuştur; Temmuz ayında en düşük birim dengesizlik maliyeti (61.90 TL/MWh) ve en düşük maliyet-gelir oranı (%2.55) kaydedilmiştir. Bu dönemde rüzgar tahminlerinin daha güvenilir olması ve yüksek kapasite faktörleri (Nisan %54.5, Temmuz %46.4) santralın optimal performans göstermesini sağlamıştır.
Kış döneminde ise ciddi zorluklar yaşanmıştır. Özellikle Aralık (-10.88%) ve Ocak (-8.96%) aylarında büyük eksik üretimler gerçekleşirken, Şubat ayında tam tersi bir durum yaşanarak %16.28 fazla üretim kaydedilmiştir. Bu durum kış aylarında rüzgar tahmin modellerinin yetersizliğini ve meteorolojik belirsizliklerin arttığını göstermektedir.
En kritik sorun Eylül ayında görülmüştür: %18.7 ile en düşük kapasite faktörü, 160.42 TL/MWh ile en yüksek birim dengesizlik maliyeti ve %7.08 ile en yüksek maliyet-gelir oranlarından biri. Bu durum düşük rüzgar potansiyeli olan dönemlerde tahmin yapmanın zorluğunu ve mali riskini açıkça ortaya koymaktadır. Genel olarak santral yılda 20 milyon TL dengesizlik maliyeti yaşamış ve ortalama %5.23 maliyet-gelir oranı ile çalışmıştır. Rüzgar tahmin sistemlerinin özellikle geçiş dönemlerinde (Eylül-Ekim) geliştirilmesi kritik öneme sahiptir.

In [20]:
df_monthly["DM_Gelir_Orani"].mean()

np.float64(5.169641666666666)

In [9]:
print("=== MASLAKTEPE RES - Risk Analizi ===")

# En yüksek DM yaşanan günler
daily_dm = df.groupby('Tarih')['Dengesizlik_Maliyeti'].sum()
print("En yüksek dengesizlik maliyeti yaşanan 10 gün:")
# En yüksek dengesizlik maliyeti yaşanan 10 günü, önce DM'ye göre en yüksekten seçip, sonra tarihe göre (en yeni gün en üstte) göstermek için:
top10_dm_days = daily_dm.sort_values(ascending=False).head(10)
top10_dm_days = top10_dm_days.sort_index(ascending=False)
print(top10_dm_days)

# En yüksek maliyetli günlerin toplam istatistikleri
print(f"\nEn Maliyetli 10 Gün - Genel İstatistikler:")
print(f"Toplam Maliyet: {top10_dm_days.sum():,.0f} TL")
print(f"Ortalama Maliyet: {top10_dm_days.mean():,.0f} TL/gün")
print(f"Toplam Maliyetteki Payı: {top10_dm_days.sum() / df['Dengesizlik_Maliyeti'].sum() * 100:.1f}%")

hourly_analysis = df.groupby('Saat').agg({
    'Dengesizlik_Maliyeti': 'mean',
    'Dengesizlik': 'mean',
    'PTF': 'mean',
    'KGUP': 'mean'
}).reset_index()

# Dengesizlik oranı = dengesizlik / planlanan üretim
hourly_analysis['Dengesizlik_Orani'] = (
    hourly_analysis['Dengesizlik'] / hourly_analysis['KGUP'] * 100
).round(2)

# En riskli saatler
risky_hours = hourly_analysis.nlargest(5, 'Dengesizlik_Orani')[['Dengesizlik_Maliyeti', 'Dengesizlik_Orani', 'PTF']]
print(f"\nEn riskli 5 saat:")
print(risky_hours)

"""risky_hours = hourly_analysis.nlargest(5, 'Dengesizlik_Maliyeti')[['Dengesizlik_Maliyeti', 'Dengesizlik_Orani', 'PTF']]
print(f"\nEn riskli 5 saat:")
print(risky_hours)"""

# Mevsimsel risk analizi
seasonal_risk = df.groupby('Ay').agg({
    'Dengesizlik_Maliyeti': ['sum', 'mean', 'std'],
    'Birim_DM': 'mean'
}).round(2)
seasonal_risk.columns = ['Toplam_DM', 'Ortalama_DM', 'DM_Volatilite', 'Birim_DM']
seasonal_risk = seasonal_risk.sort_values('Birim_DM', ascending=False)
print(f"\nMevsimsel Risk Profili (Birim_DM'ye göre büyükten küçüğe sıralı):")
print(seasonal_risk)

# Extreme event analizi
extreme_dm = df[df['Dengesizlik_Maliyeti'] > df['Dengesizlik_Maliyeti'].quantile(0.95)]
print(f"\nExtreme DM Events (>95. persentil):")
print(f"Toplam olay sayısı: {len(extreme_dm)}")
print(f"Bu olayların ortalama PTF'si: {extreme_dm['PTF'].mean():.0f} TL/MWh")
print(f"Bu olayların ortalama dengesizlik büyüklüğü: {extreme_dm['Dengesizlik'].abs().mean():.2f} MWh")

=== MASLAKTEPE RES - Risk Analizi ===
En yüksek dengesizlik maliyeti yaşanan 10 gün:
Tarih
2024-11-21    254610.384246
2024-10-28    383524.690790
2024-07-06    242691.065646
2024-04-29    258908.826085
2024-04-17    329361.085419
2024-04-16    270270.322513
2024-04-13    308518.818945
2024-01-30    257070.184448
2024-01-22    314438.341702
2024-01-14    250818.428343
Name: Dengesizlik_Maliyeti, dtype: float64

En Maliyetli 10 Gün - Genel İstatistikler:
Toplam Maliyet: 2,870,212 TL
Ortalama Maliyet: 287,021 TL/gün
Toplam Maliyetteki Payı: 13.7%

En riskli 5 saat:
    Dengesizlik_Maliyeti  Dengesizlik_Orani          PTF
14           2876.671378               8.95  2082.846885
13           3142.955312               8.16  1750.073661
11           2682.402135               6.38  1997.842404
12           2769.577944               6.36  1503.278552
15           2749.550915               4.12  2164.695464

Mevsimsel Risk Profili (Birim_DM'ye göre büyükten küçüğe sıralı):
     Toplam_DM  Ortal

In [23]:
# En maliyetli 30 saat detay analizi
print("=== MASLAKTEPE RES - En Maliyetli 30 Saat Detay Analizi ===")

# En yüksek dengesizlik maliyeti olan 30 saati seç
top_30_cost = df.nlargest(30, 'Dengesizlik_Maliyeti').copy()
top_30_cost['PTF_SMF_Spread'] = abs(top_30_cost['PTF'] - top_30_cost['SMF'])

print(f"En Maliyetli 30 Saat - Genel İstatistikler:")
print(f"Toplam Maliyet: {top_30_cost['Dengesizlik_Maliyeti'].sum():,.0f} TL")
print(f"Ortalama Maliyet: {top_30_cost['Dengesizlik_Maliyeti'].mean():,.0f} TL/saat")
print(f"Toplam Maliyetteki Payı: {top_30_cost['Dengesizlik_Maliyeti'].sum() / df['Dengesizlik_Maliyeti'].sum() * 100:.1f}%")

# Zaman dağılımı
time_distribution = top_30_cost.groupby('Saat').size().sort_values(ascending=False)
print(f"\nEn Maliyetli Saatlerin Zaman Dağılımı:")
print(time_distribution.head())

month_distribution = top_30_cost.groupby('Ay').size().sort_values(ascending=False)
print(f"\nEn Maliyetli Saatlerin Ay Dağılımı:")
print(month_distribution.head())

# Ortak karakteristikler
print(f"\nEn Maliyetli 30 Saatin Karakteristikleri:")
characteristics = top_30_cost.agg({
    'PTF': ['mean', 'min', 'max'],
    'SMF': ['mean', 'min', 'max'],
    'URETIM': ['mean', 'min', 'max'],
    'KGUP': ['mean', 'min', 'max'],
    'Dengesizlik': ['mean', 'min', 'max'],
    'PTF_SMF_Spread': ['mean', 'min', 'max'] if 'PTF_SMF_Spread' in df.columns else ['mean']
}).round(2)

print(characteristics)

# Bu saatlerdeki dengesizlik yönü
positive_imbalance = (top_30_cost['Dengesizlik'] > 0).sum()
negative_imbalance = (top_30_cost['Dengesizlik'] < 0).sum()
print(f"\nDengesizlik Yönü Dağılımı:")
print(f"Pozitif (Fazla Üretim): {positive_imbalance} saat")
print(f"Negatif (Az Üretim): {negative_imbalance} saat")

# Detaylı liste (ilk 10 saat)
print(f"\nEn Maliyetli 10 Saat Detayı:")
detailed_top10 = top_30_cost.head(10)[['Tarih', 'Saat', 'PTF', "SMF", 'KGUP', 'URETIM', 'Dengesizlik', 'Dengesizlik_Maliyeti']]
detailed_top10

=== MASLAKTEPE RES - En Maliyetli 30 Saat Detay Analizi ===
En Maliyetli 30 Saat - Genel İstatistikler:
Toplam Maliyet: 1,543,595 TL
Ortalama Maliyet: 51,453 TL/saat
Toplam Maliyetteki Payı: 7.4%

En Maliyetli Saatlerin Zaman Dağılımı:
Saat
11:00    3
06:00    3
00:00    2
21:00    2
15:00    2
dtype: int64

En Maliyetli Saatlerin Ay Dağılımı:
Ay
4     8
10    6
11    5
1     4
7     3
dtype: int64

En Maliyetli 30 Saatin Karakteristikleri:
          PTF      SMF  URETIM   KGUP  Dengesizlik  PTF_SMF_Spread
mean  1969.43  1357.72   36.62  32.23         4.40         1513.95
min    500.00     0.19    0.00   5.50       -61.70             NaN
max   2850.00  2999.05   62.00  62.40        45.11             NaN

Dengesizlik Yönü Dağılımı:
Pozitif (Fazla Üretim): 19 saat
Negatif (Az Üretim): 11 saat

En Maliyetli 10 Saat Detayı:


Tarih   Saat      PTF      SMF   KGUP  URETIM  Dengesizlik  \
4134 2024-06-21  06:00  1248.00  2600.00  61.70    0.00       -61.70   
4503 2024-07-06  15:00  1200.00  2998.40  62.40   22.97       -39.43   
2857 2024-04-29  01:00  2700.00     5.00   9.70   37.00        27.30   
7245 2024-10-28  21:00  2799.99   710.00  29.20   61.00        31.80   
2478 2024-04-13  06:00  1445.98     0.50  15.65   60.76        45.11   
2547 2024-04-16  03:00  2145.08   750.00  18.95   62.00        43.05   
2581 2024-04-17  13:00   500.00  1725.01  55.00    9.00       -46.00   
7812 2024-11-21  12:00  1415.15  2999.05  62.40   27.35       -35.05   
6731 2024-10-07  11:00  2316.34  1051.00  17.10   62.00        44.90   
7246 2024-10-28  22:00  2626.26   675.00  32.00   61.28        29.28   

      Dengesizlik_Maliyeti  
4134          88231.000000  
4503          74457.719360  
2857          73577.595000  
7245          67139.022000  
2478          65206.279450  
2547          61026.819000  
2581          58730.973800  
7812          58669.196075  
6731          58229.463000  
7246          57725.812800

In [36]:
STEP = 7.0

def optimize_kgnp_combined(df, risk_aylar, risk_saatler, maliyet_aylar, maliyet_saatler):
    """
    Kesişen zamanlar için risk stratejisi, kalanlar için maliyet stratejisi uygular
    """
    df_opt = df.copy()
    
    if df_opt['Saat'].dtype == 'object':
        df_opt['Saat'] = pd.to_datetime(df_opt['Saat'], format='%H:%M').dt.hour
    
    df_opt['KGUP_Opt'] = df_opt['KGUP']
    df_opt['Strateji'] = 'Değişiklik_Yok'
    
    # Risk ve maliyet kümelerini tanımla
    risk_set = set((ay, saat) for ay in risk_aylar for saat in risk_saatler)
    maliyet_set = set((ay, saat) for ay in maliyet_aylar for saat in maliyet_saatler)
    
    # Kesişenleri bul
    kesisen_set = risk_set.intersection(maliyet_set)
    
    # Sadece risk setinde olanlar (kesişenler hariç)
    sadece_risk_set = risk_set - kesisen_set
    
    # Sadece maliyet setinde olanlar (kesişenler hariç)
    sadece_maliyet_set = maliyet_set - kesisen_set
    
    print(f"Kesişen zaman dilimleri: {len(kesisen_set)}")
    print(f"Sadece risk stratejisi: {len(sadece_risk_set)}")
    print(f"Sadece maliyet stratejisi: {len(sadece_maliyet_set)}")
    print(f"Toplam optimize edilecek kombinasyon: {len(kesisen_set) + len(sadece_risk_set) + len(sadece_maliyet_set)}")
    
    # Optimizasyon uygula
    for idx, row in df_opt.iterrows():
        ay_saat = (row['Ay'], row['Saat'])
        
        # Strateji uygulanacak mı kontrol et
        if ay_saat in kesisen_set or ay_saat in sadece_risk_set or ay_saat in sadece_maliyet_set:
            
            # Strateji adını belirle
            if ay_saat in kesisen_set or ay_saat in sadece_risk_set:
                 df_opt.at[idx, 'Strateji'] = 'Risk_Stratejisi'
            else: # sadece_maliyet_set
                 df_opt.at[idx, 'Strateji'] = 'Maliyet_Stratejisi'

            # Ortak optimizasyon mantığını uygula
            kgup = row['KGUP_Opt']
            uretim = row['URETIM']
            ptf = row['PTF']
            poz = row['Poz_DF']
            neg = row['Neg_DF']
            
            d = uretim - kgup
            
            if d < 0:  # eksik üretim
                if neg > ptf:
                    kgup -= STEP
            else:      # fazla üretim
                if poz < ptf:
                    kgup += STEP
            
            if kgup < 0:
                kgup = 0
            df_opt.at[idx, 'KGUP_Opt'] = kgup
    
    # Gelir hesaplamaları
    df_opt['GOP_Opt'] = df_opt['KGUP_Opt'] * df_opt['PTF']
    df_opt['Dengesizlik_Opt'] = df_opt['URETIM'] - df_opt['KGUP_Opt']
    df_opt['Fiyat_Opt'] = df_opt.apply(
        lambda row: row['Poz_DF'] if row['Dengesizlik_Opt'] >= 0 else row['Neg_DF'], axis=1
    )
    df_opt['Dengesizlik_Tutarı_Opt'] = df_opt['Dengesizlik_Opt'] * df_opt['Fiyat_Opt']
    df_opt['Net_Opt'] = df_opt['GOP_Opt'] + df_opt['Dengesizlik_Tutarı_Opt']
    
    # Sonuçlar
    toplam_mevcut = df['Net_Gelir'].sum()
    toplam_yeni = df_opt['Net_Opt'].sum()
    iyilestirme = toplam_yeni - toplam_mevcut
    
    print(f"\n=== SONUÇLAR ===")
    print(f"Mevcut Net Gelir: {toplam_mevcut:,.0f} TL")
    print(f"Optimize Net Gelir: {toplam_yeni:,.0f} TL")
    print(f"Toplam İyileştirme: {iyilestirme:,.0f} TL")
    
    # Strateji bazında detay
    strateji_detay = df_opt.groupby('Strateji').agg({
        'Net_Gelir': 'sum',
        'Net_Opt': 'sum'
    })
    strateji_detay['Iyilestirme'] = strateji_detay['Net_Opt'] - strateji_detay['Net_Gelir']
    
    print(f"\n=== STRATEJİ BAZINDA DETAY ===")
    print(strateji_detay)
    
    return df_opt

df_opt_combined = optimize_kgnp_combined(
    df, 
    risk_aylar=[1, 3, 6, 9, 11, 12], 
    risk_saatler=[11, 12, 13, 14, 15],
    maliyet_aylar=[1, 4, 7, 10, 11], 
    maliyet_saatler=[0, 6, 11, 15, 21]
)

Kesişen zaman dilimleri: 4
Sadece risk stratejisi: 26
Sadece maliyet stratejisi: 21
Toplam optimize edilecek kombinasyon: 51

=== SONUÇLAR ===
Mevcut Net Gelir: 423,914,448 TL
Optimize Net Gelir: 425,247,823 TL
Toplam İyileştirme: 1,333,375 TL

=== STRATEJİ BAZINDA DETAY ===
                       Net_Gelir       Net_Opt    Iyilestirme
Strateji                                                     
Değişiklik_Yok      3.491659e+08  3.491659e+08       0.000000
Maliyet_Stratejisi  3.952392e+07  4.017392e+07  649995.362370
Risk_Stratejisi     3.522461e+07  3.590799e+07  683379.679911


In [37]:
STEP = 7.0

def optimize_kgnp(df, riskli_aylar, riskli_saatler, use_mean=True):
    """
    Belirlenen ay ve saatlere göre KGUP optimizasyonu yapar
    ve özet rapor döner.
    """
    df_opt = df.copy()

    # Saat string → int
    if df_opt['Saat'].dtype == 'object':
        df_opt['Saat'] = pd.to_datetime(df_opt['Saat'], format='%H:%M').dt.hour

    df_opt['KGUP_Opt'] = df_opt['KGUP']

    # --- optimizasyon ---
    for idx, row in df_opt.iterrows():
        if (row['Saat'] in riskli_saatler) and (row['Ay'] in riskli_aylar):
            kgup = row['KGUP_Opt']
            uretim = row['URETIM']
            ptf = row['PTF']
            poz = row['Poz_DF']
            neg = row['Neg_DF']

            d = uretim - kgup  # dengesizlik

            if d < 0:  # eksik üretim
                if neg > ptf:
                    kgup -= STEP
            else:      # fazla üretim
                if poz < ptf:
                    kgup += STEP

            if kgup < 0:
                kgup = 0

            df_opt.at[idx, 'KGUP_Opt'] = kgup

    # --- gelir hesaplamaları ---
    df_opt['GOP_Opt'] = df_opt['KGUP_Opt'] * df_opt['PTF']
    df_opt['Dengesizlik_Opt'] = df_opt['URETIM'] - df_opt['KGUP_Opt']
    df_opt['Fiyat_Opt'] = df_opt.apply(
        lambda row: row['Poz_DF'] if row['Dengesizlik_Opt'] >= 0 else row['Neg_DF'], axis=1
    )
    df_opt['Dengesizlik_Tutarı_Opt'] = df_opt['Dengesizlik_Opt'] * df_opt['Fiyat_Opt']
    df_opt['Net_Opt'] = df_opt['GOP_Opt'] + df_opt['Dengesizlik_Tutarı_Opt']

    # --- toplam sonuç ---
    toplam_mevcut = df['Net_Gelir'].sum()
    toplam_yeni = df_opt['Net_Opt'].sum()
    iyilestirme = toplam_yeni - toplam_mevcut

    print("Mevcut Net Gelir:", toplam_mevcut)
    print("Optimize Net Gelir:", toplam_yeni)
    print("İyileştirme:", iyilestirme)

    # --- detay rapor ---
    ay_ad_tr = {
        1: 'ocak', 2: 'şubat', 3: 'mart', 4: 'nisan', 5: 'mayıs', 6: 'haziran',
        7: 'temmuz', 8: 'ağustos', 9: 'eylül', 10: 'ekim', 11: 'kasım', 12: 'aralık'
    }

    agg_fn = np.mean if use_mean else np.sum
    label = "ort" if use_mean else "top"

    for ay in riskli_aylar:
        for saat in riskli_saatler:
            mask = (df['Ay'] == ay) & (df['Saat'] == saat)
            if mask.sum() == 0:
                continue

            kgup_mevcut = agg_fn(df.loc[mask, 'KGUP'])
            uretim_mevcut = agg_fn(df.loc[mask, 'URETIM'])
            kgup_yeni = agg_fn(df_opt.loc[mask, 'KGUP_Opt'])
            net_mevcut = agg_fn(df.loc[mask, 'Net_Gelir'])
            net_yeni = agg_fn(df_opt.loc[mask, 'Net_Opt'])
            delta_kar = net_yeni - net_mevcut

            print(
                f"ay: {ay_ad_tr.get(ay, ay)} saat: {saat:02d}:00 "
                f"mevcut KGUP({label}): {kgup_mevcut:.2f} "
                f"mevcut uretim({label}): {uretim_mevcut:.2f} "
                f"yeni KGUP({label}): {kgup_yeni:.2f} "
                f"kar_yeni({label}): {net_yeni:,.0f} "
                f"kar_artis({label}): {delta_kar:,.0f}"
            )

    return df_opt

In [56]:
df_opt_risk = optimize_kgnp(df, riskli_aylar=[1,4,7,10,11], riskli_saatler=[11,12,13,14,15]) # riskli saatlerde KGUP optimizasyonu

Mevcut Net Gelir: 423914447.51297605
Optimize Net Gelir: 424790439.819883
İyileştirme: 875992.3069069386


In [39]:
df_opt_maliyet = optimize_kgnp(df, riskli_aylar=[1,4,7,10,11], riskli_saatler=[0,6,11,15,21]) # maliyetli saatlerde KGUP optimizasyonu

Mevcut Net Gelir: 423914447.51297605
Optimize Net Gelir: 424679743.03554
İyileştirme: 765295.5225639343


In [12]:
df.head()

Tarih  Ay   Saat      PTF      SMF     Poz_DF     Neg_DF  KGUP  \
0  2024-01-01   1  00:00  1299.98   560.00   543.2000  1338.9794  3.44   
1  2024-01-01   1  01:00  1299.98  1299.98  1260.9806  1338.9794  5.08   
2  2024-01-01   1  02:00  1248.54  1248.54  1211.0838  1285.9962  6.16   
3  2024-01-01   1  03:00  1299.98  1250.00  1212.5000  1338.9794  7.08   
4  2024-01-01   1  04:00  1200.00  1150.00  1115.5000  1236.0000  8.89   

   URETIM  Dengesizlik  GOP_Geliri  Dengesizlik_Tutarı     Net_Gelir  \
0    14.0        10.56   4471.9312         5736.192000  10208.123200   
1    19.0        13.92   6603.8984        17552.849952  24156.748352   
2    28.0        21.84   7691.0064        26450.070192  34141.076592   
3    32.0        24.92   9203.8584        30215.500000  39419.358400   
4    24.0        15.11  10668.0000        16855.205000  27523.205000   

   Dengesizlik_Maliyeti    Birim_DM  
0           7991.596800  570.828343  
1            542.871648   28.572192  
2            818.043408   29.215836  
3           2180.001600   68.125050  
4           1276.795000   53.199792

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Tarih                 8784 non-null   object 
 1   Ay                    8784 non-null   int64  
 2   Saat                  8784 non-null   object 
 3   PTF                   8784 non-null   float64
 4   SMF                   8784 non-null   float64
 5   Poz_DF                8784 non-null   float64
 6   Neg_DF                8784 non-null   float64
 7   KGUP                  8784 non-null   float64
 8   URETIM                8784 non-null   float64
 9   Dengesizlik           8784 non-null   float64
 10  GOP_Geliri            8784 non-null   float64
 11  Dengesizlik_Tutarı    8784 non-null   float64
 12  Net_Gelir             8784 non-null   float64
 13  Dengesizlik_Maliyeti  8784 non-null   float64
 14  Birim_DM              8784 non-null   float64
dtypes: float64(12), int64

In [13]:
# Saat sütununu integer'a çevir
df['Saat_int'] = df['Saat'].str[:2].astype(int)

risk_aylar = [1, 3, 6, 9, 11, 12]
risk_saatler = [11, 12, 13, 14, 15]

riskli_zamanlar = df[df["Ay"].isin(risk_aylar) & df["Saat_int"].isin(risk_saatler)]
print(f"Riskli zamanlardaki satır sayısı: {len(riskli_zamanlar)}")
riskli_zamanlar.head()

Riskli zamanlardaki satır sayısı: 915


Tarih  Ay   Saat      PTF      SMF     Poz_DF     Neg_DF   KGUP  \
11  2024-01-01   1  11:00  1200.00     1.19     1.1543  1236.0000  20.85   
12  2024-01-01   1  12:00  1000.00     1.88     1.8236  1030.0000  23.72   
13  2024-01-01   1  13:00  1199.99   500.00   485.0000  1235.9897  27.60   
14  2024-01-01   1  14:00  1299.98  1299.98  1260.9806  1338.9794  30.63   
15  2024-01-01   1  15:00  1299.98  1299.98  1260.9806  1338.9794  31.34   

    URETIM  Dengesizlik  GOP_Geliri  Dengesizlik_Tutarı     Net_Gelir  \
11    34.0        13.15  25020.0000           15.179045  25035.179045   
12    33.0         9.28  23720.0000           16.923008  23736.923008   
13    35.0         7.40  33119.7240         3589.000000  36708.724000   
14    50.0        19.37  39818.3874        24425.194222  64243.581622   
15    47.0        15.66  40741.3732        19746.956196  60488.329396   

    Dengesizlik_Maliyeti    Birim_DM  Saat_int  
11          15764.820955  463.671205        11  
12           9263.076992  280.699303        12  
13           5290.926000  151.169314        13  
14            755.418378   15.108368        14  
15            610.730604   12.994268        15

In [18]:
# Saat sütununu integer'a çevir
df['Saat_int'] = df['Saat'].str[:2].astype(int)

maliyet_aylar=[1, 4, 7, 10, 11]
maliyet_saatler=[0, 6, 11, 15, 21]

maliyetli_zamanlar = df[df["Ay"].isin(maliyet_aylar) & df["Saat_int"].isin(maliyet_saatler)]
print(f"Riskli zamanlardaki satır sayısı: {len(maliyetli_zamanlar)}")
maliyetli_zamanlar.sample(10)

Riskli zamanlardaki satır sayısı: 765


Tarih  Ay   Saat      PTF      SMF     Poz_DF     Neg_DF   KGUP  \
471   2024-01-20   1  15:00  2000.01  2399.99  1940.0097  2471.9897  49.60   
419   2024-01-18   1  11:00  2499.67  1700.00  1649.0000  2574.6601  49.60   
7776  2024-11-20  11  00:00  2990.00  2999.00  2900.3000  3088.9700  45.80   
6957  2024-10-16  10  21:00  2188.39  2950.00  2122.7383  3038.5000  26.00   
7704  2024-11-17  11  00:00  2200.00  2200.00  2134.0000  2266.0000  36.10   
4728  2024-07-16   7  00:00  2685.56  2685.56  2604.9932  2766.1268  39.10   
4440  2024-07-04   7  00:00  3000.00  3000.00  2910.0000  3090.0000   5.40   
4869  2024-07-21   7  21:00  3000.00  3000.00  2910.0000  3090.0000   9.50   
2325  2024-04-06   4  21:00  2620.99  2693.00  2542.3603  2773.7900  41.67   
4643  2024-07-12   7  11:00  2850.01  2999.04  2764.5097  3089.0112   6.30   

      URETIM  Dengesizlik   GOP_Geliri  Dengesizlik_Tutarı      Net_Gelir  \
471    50.00         0.40   99200.4960          776.003880   99976.499880   
419    50.00         0.40  123983.6320          659.600000  124643.232000   
7776   40.28        -5.52  136942.0000       -17051.114400  119890.885600   
6957   20.20        -5.80   56898.1400       -17623.300000   39274.840000   
7704   19.24       -16.86   79420.0000       -38204.760000   41215.240000   
4728   28.00       -11.10  105005.3960       -30704.007480   74301.388520   
4440    9.71         4.31   16200.0000        12542.100000   28742.100000   
4869    8.37        -1.13   28500.0000        -3491.700000   25008.300000   
2325   40.79        -0.88  109216.6533        -2440.935200  106775.718100   
4643    3.62        -2.68   17955.0630        -8278.550016    9676.512984   

      Dengesizlik_Maliyeti    Birim_DM  Saat_int  
471              24.000120    0.480002        15  
419             340.268000    6.805360        11  
7776            546.314400   13.562920         0  
6957           4930.638000  244.090990        21  
7704           1112.760000   57.835759         0  
4728            894.291480   31.938981         0  
4440            387.900000   39.948507         0  
4869            101.700000   12.150538        21  
2325            134.464000    3.296494        21  
4643            640.523216  176.940115        11

In [40]:
def kpi_summary(df_monthly, plant_name="Santral"):
    num_cols = df_monthly.select_dtypes(include='number').columns

    summary = pd.DataFrame({
        "SUM": df_monthly[num_cols].sum(),
        "MEAN": df_monthly[num_cols].mean()
    }).round(2)

    summary.index.name = f"{plant_name} - KPI"
    return summary

eber_summary = kpi_summary(df_monthly, "EBER_RES")
eber_summary

SUM         MEAN
EBER_RES - KPI                                 
KGUP                  2.033220e+05     16943.50
URETIM                2.037793e+05     16981.61
Dengesizlik           4.572700e+02        38.11
GOP_Geliri            4.440038e+08  37000316.62
Dengesizlik_Tutarı   -2.008935e+07  -1674112.66
Net_Gelir             4.239144e+08  35326203.96
Dengesizlik_Maliyeti  2.090187e+07   1741822.31
Dengesizlik_Orani     1.458000e+01         1.21
Birim_DM_Aylik        1.265490e+03       105.46
Kapasite_Faktoru      4.455000e+02        37.12
DM_Gelir_Orani        6.204000e+01         5.17

Yorum: Gerçek üretim, planlanandan çok az fazla. Dengesizlik oranı %1.21 ile oldukça düşük, tahminler başarılı.
Her üretilen MWh başına ortalama 105 TL dengesizlik maliyeti ödenmiş. Maslaktepe’ye göre daha yüksek, bu da Eber’de dengesizliklerin daha pahalıya mal olduğunu gösteriyor.
Dengesizlik maliyeti Maslaktepe’den yüksek: Toplam gelirin %5.17’si seviyesinde
%37.1 kapasite faktörü, santralin teknik ve operasyonel olarak verimli çalıştığını gösteriyor, ancak Maslaktepe’ye göre biraz daha düşük.

In [41]:
# Dengesizlik performans analizi
print("=== EBER RES - Dengesizlik Performans Analizi ===")
print(f"Ortalama Dengesizlik Oranı: {(df['Dengesizlik'].sum() / df['KGUP'].sum() * 100):.2f}%")
print(f"Pozitif Dengesizlik Oranı: {(df['Dengesizlik'] > 0).sum() / len(df) * 100:.1f}%")
print(f"Negatif Dengesizlik Oranı: {(df['Dengesizlik'] < 0).sum() / len(df) * 100:.1f}%")
print(f"Sıfır Üretim Saatleri: {(df['URETIM'] == 0).sum()} saat ({(df['URETIM'] == 0).sum()/len(df)*100:.1f}%)")

# Dengesizlik büyüklük dağılımı
dengesizlik_abs = df['Dengesizlik'].abs()
print(f"\nDengesizlik Büyüklük İstatistikleri:")
print(f"Medyan: {dengesizlik_abs.median():.3f} MWh")
print(f"Ortalama: {dengesizlik_abs.mean():.3f} MWh") 
print(f"95. Persentil: {dengesizlik_abs.quantile(0.95):.3f} MWh")
print(f"Maksimum: {dengesizlik_abs.max():.3f} MWh")

=== EBER RES - Dengesizlik Performans Analizi ===
Ortalama Dengesizlik Oranı: 0.22%
Pozitif Dengesizlik Oranı: 50.1%
Negatif Dengesizlik Oranı: 49.6%
Sıfır Üretim Saatleri: 751 saat (8.6%)

Dengesizlik Büyüklük İstatistikleri:
Medyan: 4.610 MWh
Ortalama: 7.490 MWh
95. Persentil: 24.698 MWh
Maksimum: 62.400 MWh


In [42]:
def performance_nonzero(df, name):
    # Sadece üretim yapılan saatleri al
    active = df[df["URETIM"] > 0].copy()
    result = {"Santral": name, "Aktif_Saat": len(active)}

    # Tahmine göre oransal sapma
    valid = active[active["KGUP"] > 0]
    result["Ortalama_Dengesizlik_%"] = (valid["Dengesizlik"].abs() / valid["KGUP"]).mean() * 100
    result["Medyan_Dengesizlik_%"]   = (valid["Dengesizlik"].abs() / valid["KGUP"]).median() * 100

    # Ortalama bias (pozitif = fazla üretim)
    result["Bias_MWh"]     = active["Dengesizlik"].mean()
    result["Pozitif_%"]    = (active["Dengesizlik"] > 0).mean() * 100
    result["Negatif_%"]    = (active["Dengesizlik"] < 0).mean() * 100

    # Dengesizlik büyüklüğü istatistikleri
    result["Medyan_MWh"]   = active["Dengesizlik"].median()
    result["Ortalama_MWh"] = active["Dengesizlik"].mean()
    result["P95_MWh"]      = active["Dengesizlik"].quantile(0.95)
    result["Max_MWh"]      = active["Dengesizlik"].max()

    return result

eber_perf  = performance_nonzero(df, "EBER_RES")

pd.DataFrame([eber_perf]).round(3)

Santral  Aktif_Saat  Ortalama_Dengesizlik_%  Medyan_Dengesizlik_%  \
0  EBER_RES        8031                 107.226                  37.5   

   Bias_MWh  Pozitif_%  Negatif_%  Medyan_MWh  Ortalama_MWh  P95_MWh  Max_MWh  
0     0.441     54.788     45.026        0.53         0.441     19.0    49.89

In [43]:
print("=== EBER RES - Finansal Etki Analizi ===")
total_net_revenue = df['Net_Gelir'].sum()
total_dm_cost = df['Dengesizlik_Maliyeti'].sum()
total_production = df['URETIM'].sum()

print(f"Toplam Net Gelir: {total_net_revenue:,.0f} TL")
print(f"Toplam Dengesizlik Maliyeti: {total_dm_cost:,.0f} TL")
print(f"DM'nin Net Gelir'deki Payı: {total_dm_cost/total_net_revenue*100:.2f}%")
print(f"Yıllık Birim Dengesizlik Maliyeti: {total_dm_cost/total_production:.2f} TL/MWh")

# Aylık finansal performans
monthly_financial = df_monthly[['Net_Gelir', 'Dengesizlik_Maliyeti', 'DM_Gelir_Orani']].copy()
print(f"\nEn yüksek DM/Gelir oranına sahip aylar:")
print(monthly_financial.nlargest(3, 'DM_Gelir_Orani'))

# PTF-SMF spread analizi
df['PTF_SMF_Spread'] = abs(df['PTF'] - df['SMF'])
spread_analysis = df.groupby(pd.cut(df['PTF_SMF_Spread'], bins=5)).agg({
    'Dengesizlik_Maliyeti': 'mean',
    'Birim_DM': 'mean'
}).round(2)
print(f"\nPTF-SMF Spread'e göre DM analizi:")
spread_analysis

=== EBER RES - Finansal Etki Analizi ===
Toplam Net Gelir: 423,914,448 TL
Toplam Dengesizlik Maliyeti: 20,901,868 TL
DM'nin Net Gelir'deki Payı: 4.93%
Yıllık Birim Dengesizlik Maliyeti: 102.57 TL/MWh

En yüksek DM/Gelir oranına sahip aylar:
      Net_Gelir  Dengesizlik_Maliyeti  DM_Gelir_Orani
Ay                                                   
9   19082737.37            1350620.09          7.0777
1   35917848.07            2498113.95          6.9551
4   41062006.27            2762033.92          6.7265

PTF-SMF Spread'e göre DM analizi:


/var/folders/1t/15hmc_h12xnbygjnmf7kj43h0000gn/T/ipykernel_1820/905342762.py:19: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  spread_analysis = df.groupby(pd.cut(df['PTF_SMF_Spread'], bins=5)).agg({


Dengesizlik_Maliyeti  Birim_DM
PTF_SMF_Spread                                  
(-2.695, 539.0]                1217.65    161.62
(539.0, 1078.0]                4311.71    422.32
(1078.0, 1617.0]               6395.37    432.25
(1617.0, 2156.0]               9909.04    537.67
(2156.0, 2695.0]               7593.41    327.44

In [45]:
def eber_dengesizlik_yonu_analizi(df):
    """
    EBER RES için pozitif vs negatif dengesizlik analizi
    Hangi yönde hata yapmak daha pahalı?
    """
    
    print("=" * 70)
    print("EBER RES - POZİTİF vs NEGATİF DENGESİZLİK ANALİZİ")
    print("=" * 70)
    
    # Dengesizlik yönlerini kategorize et
    df_analysis = df.copy()
    df_analysis['dengesizlik_yonu'] = 'Nötr'
    df_analysis.loc[df_analysis['Dengesizlik'] > 0, 'dengesizlik_yonu'] = 'Pozitif_Dengesizlik'  # Fazla ürettik
    df_analysis.loc[df_analysis['Dengesizlik'] < 0, 'dengesizlik_yonu'] = 'Negatif_Dengesizlik'  # Az ürettik
    
    # Her yön için detaylı analiz
    yön_analizi = df_analysis.groupby('dengesizlik_yonu').agg({
        'Dengesizlik': ['count', 'mean', 'std', 'min', 'max'],
        'Dengesizlik_Maliyeti': ['sum', 'mean', 'max'],
        'Birim_DM': ['mean', 'max'],
        'PTF': 'mean',
        'SMF': 'mean',
        'Poz_DF': 'mean',
        'Neg_DF': 'mean'
    }).round(2)
    
    print("\nYÖN BAZLI İSTATİSTİKLER:")
    print("-" * 50)
    
    for yön in ['Pozitif_Dengesizlik', 'Negatif_Dengesizlik', 'Nötr']:
        if yön in yön_analizi.index:
            data = yön_analizi.loc[yön]
            saat_sayisi = data[('Dengesizlik', 'count')]
            toplam_maliyet = data[('Dengesizlik_Maliyeti', 'sum')]
            ort_maliyet = data[('Dengesizlik_Maliyeti', 'mean')]
            
            if yön == 'Pozitif_Dengesizlik':
                print(f"FAZLA ÜRETİM (Tahmin > Gerçek):")
            elif yön == 'Negatif_Dengesizlik':
                print(f"EKSIK ÜRETİM (Tahmin < Gerçek):")
            else:
                print(f"MÜKEMMEL TAHMİN:")
            
            print(f"   Olay Sayısı: {saat_sayisi:.0f} saat")
            print(f"   Toplam Maliyet: {toplam_maliyet:,.0f} TL")
            print(f"   Ortalama Maliyet: {ort_maliyet:.0f} TL/saat")
            print(f"   Ortalama PTF: {data[('PTF', 'mean')]:.0f} TL/MWh")
            print(f"   Ortalama SMF: {data[('SMF', 'mean')]:.0f} TL/MWh")
            print()
    
    # Hangi yön daha maliyetli?
    pos_toplam = df_analysis[df_analysis['dengesizlik_yonu']=='Pozitif_Dengesizlik']['Dengesizlik_Maliyeti'].sum()
    neg_toplam = df_analysis[df_analysis['dengesizlik_yonu']=='Negatif_Dengesizlik']['Dengesizlik_Maliyeti'].sum()
    
    print("KRITIK BULGULAR:")
    print("-" * 30)
    print(f"Pozitif Dengesizlik Maliyeti: {pos_toplam:,.0f} TL")
    print(f"Negatif Dengesizlik Maliyeti: {neg_toplam:,.0f} TL")
    
    if neg_toplam > pos_toplam:
        risk_orani = neg_toplam / pos_toplam
        print(f"EKSIK ÜRETİM {risk_orani:.1f}x DAHA RİSKLİ!")
        print("Çünkü: SMF genelde PTF'den yüksek + %103 ceza çarpanı")
    else:
        risk_orani = pos_toplam / neg_toplam
        print(f"FAZLA ÜRETİM {risk_orani:.1f}x DAHA RİSKLİ!")
        print("Çünkü: PTF düşük dönemlerde fazla üretim zarar")
    
    # Fiyat spread analizi
    print(f"\nFİYAT SPREAD ANALİZİ:")
    print("-" * 30)
    ptf_ort = df['PTF'].mean()
    smf_ort = df['SMF'].mean()
    spread = smf_ort - ptf_ort
    
    print(f"Ortalama PTF: {ptf_ort:.0f} TL/MWh")
    print(f"Ortalama SMF: {smf_ort:.0f} TL/MWh")
    
    if spread > 0:
        print("SMF > PTF olduğu için NEGATİF dengesizlik daha riskli")
        print("   (Eksik üretimde yüksek SMF ile ceza yiyoruz)")
    else:
        print("PTF > SMF olduğu için POZİTİF dengesizlik daha riskli")
    
    # Saatlik risk analizi
    pozitif_saatlik = df_analysis[df_analysis['dengesizlik_yonu']=='Pozitif_Dengesizlik'].groupby(
        df_analysis['Saat'].str.replace(':00', '').astype(int)
    )['Dengesizlik_Maliyeti'].sum().sort_values(ascending=False)

    negatif_saatlik = df_analysis[df_analysis['dengesizlik_yonu']=='Negatif_Dengesizlik'].groupby(
        df_analysis['Saat'].str.replace(':00', '').astype(int)
    )['Dengesizlik_Maliyeti'].sum().sort_values(ascending=False)

    print("\nPOZİTİF DENGESİZLİKTE SAATLİK MALİYETLER (TL):")
    print(pozitif_saatlik)
    print("\nNEGATİF DENGESİZLİKTE SAATLİK MALİYETLER (TL):")
    print(negatif_saatlik)
    
    negatif_saatlik = df_analysis[df_analysis['dengesizlik_yonu']=='Negatif_Dengesizlik'].groupby(
        df_analysis['Saat'].str.replace(':00', '').astype(int)
    )['Dengesizlik_Maliyeti'].sum().sort_values(ascending=False)
    
    
    return {
        'pozitif_maliyet': pos_toplam,
        'negatif_maliyet': neg_toplam,
        'risk_yonu': 'negatif' if neg_toplam > pos_toplam else 'pozitif',
        'risk_orani': max(neg_toplam, pos_toplam) / min(neg_toplam, pos_toplam)
    }

# Analizi çalıştır
dengesizlik_analizi = eber_dengesizlik_yonu_analizi(df)

EBER RES - POZİTİF vs NEGATİF DENGESİZLİK ANALİZİ

YÖN BAZLI İSTATİSTİKLER:
--------------------------------------------------
FAZLA ÜRETİM (Tahmin > Gerçek):
   Olay Sayısı: 4400 saat
   Toplam Maliyet: 11,365,066 TL
   Ortalama Maliyet: 2583 TL/saat
   Ortalama PTF: 2188 TL/MWh
   Ortalama SMF: 2073 TL/MWh

EKSIK ÜRETİM (Tahmin < Gerçek):
   Olay Sayısı: 4357 saat
   Toplam Maliyet: 9,536,802 TL
   Ortalama Maliyet: 2189 TL/saat
   Ortalama PTF: 2283 TL/MWh
   Ortalama SMF: 2286 TL/MWh

MÜKEMMEL TAHMİN:
   Olay Sayısı: 25 saat
   Toplam Maliyet: 0 TL
   Ortalama Maliyet: 0 TL/saat
   Ortalama PTF: 2379 TL/MWh
   Ortalama SMF: 2218 TL/MWh

KRITIK BULGULAR:
------------------------------
Pozitif Dengesizlik Maliyeti: 11,365,066 TL
Negatif Dengesizlik Maliyeti: 9,536,802 TL
FAZLA ÜRETİM 1.2x DAHA RİSKLİ!
Çünkü: PTF düşük dönemlerde fazla üretim zarar

FİYAT SPREAD ANALİZİ:
------------------------------
Ortalama PTF: 2235 TL/MWh
Ortalama SMF: 2179 TL/MWh
PTF > SMF olduğu için POZİTİF de

In [52]:
# Pazar dinamikleri analizi
print("=== EBER RES - Pazar Dinamikleri Analizi ===")

# Yüksek/düşük fiyat dönemleri
high_price_threshold = df['PTF'].quantile(0.8)
low_price_threshold = df['PTF'].quantile(0.2)

high_price_perf = df[df['PTF'] >= high_price_threshold].agg({
    'URETIM': 'mean',
    'Dengesizlik': 'mean',
    'Dengesizlik_Maliyeti': 'mean',
    'Birim_DM': 'mean'
}).round(3)

low_price_perf = df[df['PTF'] <= low_price_threshold].agg({
    'URETIM': 'mean', 
    'Dengesizlik': 'mean',
    'Dengesizlik_Maliyeti': 'mean',
    'Birim_DM': 'mean'
}).round(3)

print(f"Yüksek Fiyat Dönemlerinde (PTF > {high_price_threshold:.0f} TL/MWh):")
print(f"Ortalama Üretim: {high_price_perf['URETIM']:.2f} MWh")
print(f"Ortalama Dengesizlik: {high_price_perf['Dengesizlik']:.3f} MWh")
print(f"Ortalama Birim DM: {high_price_perf['Birim_DM']:.2f} TL/MWh")

print(f"\nDüşük Fiyat Dönemlerinde (PTF < {low_price_threshold:.0f} TL/MWh):")
print(f"Ortalama Üretim: {low_price_perf['URETIM']:.2f} MWh")
print(f"Ortalama Dengesizlik: {low_price_perf['Dengesizlik']:.3f} MWh")
print(f"Ortalama Birim DM: {low_price_perf['Birim_DM']:.2f} TL/MWh")

=== EBER RES - Pazar Dinamikleri Analizi ===
Yüksek Fiyat Dönemlerinde (PTF > 2743 TL/MWh):
Ortalama Üretim: 21.26 MWh
Ortalama Dengesizlik: 0.011 MWh
Ortalama Birim DM: 157.52 TL/MWh

Düşük Fiyat Dönemlerinde (PTF < 1800 TL/MWh):
Ortalama Üretim: 27.91 MWh
Ortalama Dengesizlik: -0.155 MWh
Ortalama Birim DM: 349.93 TL/MWh


In [54]:
# Kayıp oranı hesabı (mükemmel senaryoya göre)
print("=== EBER RES - Kayıp Oranı Hesabı (Mükemmel Senaryo) ===")

# Mükemmel senaryo: Hiç dengesizlik olmasaydı kazanılacak gelir
perfect_revenue = df['URETIM'] * df['PTF']
actual_revenue = df['Net_Gelir']
revenue_loss = perfect_revenue - actual_revenue

# Kayıp oranları
total_perfect_revenue = perfect_revenue.sum()
total_actual_revenue = actual_revenue.sum()
total_revenue_loss = revenue_loss.sum()

print(f"Mükemmel Senaryo Geliri: {total_perfect_revenue:,.0f} TL")
print(f"Gerçek Net Gelir: {total_actual_revenue:,.0f} TL")
print(f"Toplam Gelir Kaybı: {total_revenue_loss:,.0f} TL")
print(f"Kayıp Oranı: {(total_revenue_loss/total_perfect_revenue)*100:.2f}%")

# Aylık kayıp oranları
monthly_loss = df.groupby('Ay').agg({
    'URETIM': 'sum',
    'PTF': lambda x: (df.loc[x.index, 'URETIM'] * df.loc[x.index, 'PTF']).sum(),
    'Net_Gelir': 'sum'
}).round(2)

monthly_loss['Mükemmel_Gelir'] = df.groupby('Ay').apply(lambda x: (x['URETIM'] * x['PTF']).sum())
monthly_loss['Gelir_Kaybi'] = monthly_loss['Mükemmel_Gelir'] - monthly_loss['Net_Gelir']
monthly_loss['Kayip_Orani'] = (monthly_loss['Gelir_Kaybi'] / monthly_loss['Mükemmel_Gelir'] * 100).round(2)

print(f"\nAylık Kayıp Oranları:")
print(monthly_loss[['Mükemmel_Gelir', 'Net_Gelir', 'Gelir_Kaybi', 'Kayip_Orani']].sort_values(by='Gelir_Kaybi', ascending=False))

=== EBER RES - Kayıp Oranı Hesabı (Mükemmel Senaryo) ===
Mükemmel Senaryo Geliri: 444,816,315 TL
Gerçek Net Gelir: 423,914,448 TL
Toplam Gelir Kaybı: 20,901,868 TL
Kayıp Oranı: 4.70%

Aylık Kayıp Oranları:
    Mükemmel_Gelir    Net_Gelir   Gelir_Kaybi  Kayip_Orani
Ay                                                        
4     4.382404e+07  41062006.27  2.762034e+06         6.30
1     3.841596e+07  35917848.07  2.498114e+06         6.50
10    4.826767e+07  46015844.52  2.251826e+06         4.67
12    4.790230e+07  46169228.42  1.733076e+06         3.62
11    3.873862e+07  37024425.44  1.714192e+06         4.43
5     3.048505e+07  28889994.50  1.595060e+06         5.23
2     2.497101e+07  23409496.47  1.561512e+06         6.25
8     3.920159e+07  37734658.25  1.466936e+06         3.74
3     2.953399e+07  28165593.84  1.368401e+06         4.63
9     2.043336e+07  19082737.37  1.350620e+06         6.61
7     5.375094e+07  52416385.29  1.334558e+06         2.48
6     2.929177e+07  2802622

/var/folders/1t/15hmc_h12xnbygjnmf7kj43h0000gn/T/ipykernel_1820/2551658851.py:26: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  monthly_loss['Mükemmel_Gelir'] = df.groupby('Ay').apply(lambda x: (x['URETIM'] * x['PTF']).sum())


Eber RES: Eylül (%6.61), Ocak (%6.50) ve Nisan (%6.30) en yüksek kayıp oranlarına sahip aylardır. Özellikle Eylül'de ciddi bir zorluk yaşandığı görülüyor.

In [57]:
# Over/Under asimetrisi detaylı analizi
print("=== EBER RES - Over/Under Asimetrisi Detaylı Analizi ===")

# Pozitif (Over) ve Negatif (Under) dengesizlikleri ayır
over_production = df[df['Dengesizlik'] > 0].copy()
under_production = df[df['Dengesizlik'] < 0].copy()

print(f"Over Production Saatleri: {len(over_production)} ({len(over_production)/len(df)*100:.1f}%)")
print(f"Under Production Saatleri: {len(under_production)} ({len(under_production)/len(df)*100:.1f}%)")

# Over/Under büyüklük analizi
over_stats = {
    'Count': len(over_production),
    'Avg_Size': over_production['Dengesizlik'].mean(),
    'Median_Size': over_production['Dengesizlik'].median(),
    'Max_Size': over_production['Dengesizlik'].max(),
    'Total_Volume': over_production['Dengesizlik'].sum(),
    'Avg_Cost': over_production['Dengesizlik_Maliyeti'].mean(),
    'Total_Cost': over_production['Dengesizlik_Maliyeti'].sum()
}

under_stats = {
    'Count': len(under_production),
    'Avg_Size': under_production['Dengesizlik'].mean(),
    'Median_Size': under_production['Dengesizlik'].median(),
    'Min_Size': under_production['Dengesizlik'].min(),
    'Total_Volume': under_production['Dengesizlik'].sum(),
    'Avg_Cost': under_production['Dengesizlik_Maliyeti'].mean(),
    'Total_Cost': under_production['Dengesizlik_Maliyeti'].sum()
}

asymmetry_df = pd.DataFrame({'Over': over_stats, 'Under': under_stats}).round(3)
print(f"\nOver vs Under Asimetrisi:")
print(asymmetry_df)

# Asimetri oranları
print(f"\nAsimetri Oranları:")
print(f"Hacim Asimetrisi: {abs(over_stats['Total_Volume']) / abs(under_stats['Total_Volume']):.2f}")
print(f"Maliyet Asimetrisi: {over_stats['Total_Cost'] / abs(under_stats['Total_Cost']):.2f}")
print(f"Frekans Asimetrisi: {over_stats['Count'] / under_stats['Count']:.2f}")

=== EBER RES - Over/Under Asimetrisi Detaylı Analizi ===
Over Production Saatleri: 4400 (50.1%)
Under Production Saatleri: 4357 (49.6%)

Over vs Under Asimetrisi:
                      Over        Under
Count         4.400000e+03     4357.000
Avg_Size      7.526000e+00       -7.496
Median_Size   5.005000e+00       -4.270
Max_Size      4.989000e+01          NaN
Total_Volume  3.311622e+04   -32658.950
Avg_Cost      2.582970e+03     2188.846
Total_Cost    1.136507e+07  9536801.590
Min_Size               NaN      -62.400

Asimetri Oranları:
Hacim Asimetrisi: 1.01
Maliyet Asimetrisi: 1.19
Frekans Asimetrisi: 1.01


In [58]:
# Bias & MAPE gibi tahmin doğruluğu metrikleri
print("=== EBER RES - Bias & MAPE Tahmin Doğruluğu Metrikleri ===")

# Sadece üretim yapılan saatleri al
active_hours = df[df['KGUP'] > 0].copy()

# BIAS (Ortalama Hata)
bias = active_hours['Dengesizlik'].mean()
bias_percentage = (bias / active_hours['KGUP'].mean()) * 100

# MAPE (Mean Absolute Percentage Error)
mape = (active_hours['Dengesizlik'].abs() / active_hours['KGUP']).mean() * 100

# MAE (Mean Absolute Error)
mae = active_hours['Dengesizlik'].abs().mean()

# RMSE (Root Mean Square Error)
rmse = (active_hours['Dengesizlik'] ** 2).mean() ** 0.5

# Tahmin doğruluğu (accuracy)
accuracy = 1 - (active_hours['Dengesizlik'].abs().sum() / active_hours['KGUP'].sum())

print(f"BIAS (Ortalama Hata): {bias:.3f} MWh ({bias_percentage:.2f}%)")
print(f"MAPE (Ortalama Mutlak Yüzde Hata): {mape:.2f}%")
print(f"MAE (Ortalama Mutlak Hata): {mae:.3f} MWh")
print(f"RMSE (Kök Ortalama Kare Hata): {rmse:.3f} MWh")
print(f"Tahmin Doğruluğu: {accuracy*100:.1f}%")

# Aylık tahmin performansı
monthly_forecast = active_hours.groupby('Ay').agg({
    'Dengesizlik': ['mean', lambda x: x.abs().mean(), lambda x: (x.abs() / active_hours.loc[x.index, 'KGUP']).mean() * 100],
    'KGUP': 'mean'
}).round(3)

monthly_forecast.columns = ['BIAS', 'MAE', 'MAPE', 'KGUP_Mean']
print(f"\nAylık Tahmin Performansı:")
print(monthly_forecast)

=== EBER RES - Bias & MAPE Tahmin Doğruluğu Metrikleri ===
BIAS (Ortalama Hata): 0.052 MWh (0.22%)
MAPE (Ortalama Mutlak Yüzde Hata): 106.62%
MAE (Ortalama Mutlak Hata): 7.498 MWh
RMSE (Kök Ortalama Kare Hata): 11.146 MWh
Tahmin Doğruluğu: 67.6%

Aylık Tahmin Performansı:
     BIAS     MAE     MAPE  KGUP_Mean
Ay                                   
1  -2.703   9.125   67.050     30.179
2   2.720   6.251  171.085     16.711
3  -0.186   7.344  106.216     18.347
4   1.166  10.006   72.366     32.839
5   2.200   8.297  213.496     17.071
6  -1.168   6.995  119.982     21.000
7   0.121   7.075   71.500     28.858
8   1.514   6.664  105.929     19.241
9  -0.247   5.082   78.955     11.956
10  2.135   7.514   80.396     26.123
11 -1.500   6.267  119.890     24.483
12 -3.363   9.242   75.157     30.911


In [60]:
print("=== EBER RES - RÜZGARLI Santral Fiyat Kullanım Endeksi ===")

# Fiyat seviyelerine göre üretim verimliliği (geçici değişkenler kullan)
price_quintiles = df['PTF'].quantile([0.2, 0.4, 0.6, 0.8])
price_levels = pd.cut(df['PTF'], 
                     bins=[0, price_quintiles[0.2], price_quintiles[0.4], price_quintiles[0.6], price_quintiles[0.8], float('inf')],
                     labels=['Çok_Düşük', 'Düşük', 'Orta', 'Yüksek', 'Çok_Yüksek'])

# Her fiyat seviyesinde üretim davranışı
price_efficiency = df.groupby(price_levels).agg({
    'PTF': 'mean',
    'URETIM': ['mean', 'sum', 'count'],
    'KGUP': ['mean', 'sum'],
    'Dengesizlik_Maliyeti': ['mean', 'sum']
}).round(2)

price_efficiency.columns = ['Avg_PTF', 'Avg_Production', 'Total_Production', 'Hour_Count', 'Avg_KGUP', 'Total_KGUP', 'Avg_DM', 'Total_DM']

# Fiyat Kullanım Endeksi hesapla
total_production = df['URETIM'].sum()
weighted_price = (price_efficiency['Total_Production'] * price_efficiency['Avg_PTF']).sum() / total_production
average_price = df['PTF'].mean()
price_utilization_index = weighted_price / average_price

print(f"Fiyat Seviyelerine Göre Üretim Dağılımı:")
print(price_efficiency[['Avg_PTF', 'Total_Production', 'Hour_Count', 'Avg_DM']])

print(f"\nFiyat Kullanım Endeksi: {price_utilization_index:.3f}")
print(f"(1.0'dan yüksek = yüksek fiyatlarda daha çok üretim)")

# Su rejimi stratejisi analizi
daily_production = df.groupby('Tarih').agg({
    'URETIM': 'sum',
    'PTF': 'mean',
    'Dengesizlik_Maliyeti': 'sum'
}).reset_index()

# Günlük üretim vs fiyat korelasyonu
daily_corr = daily_production['URETIM'].corr(daily_production['PTF'])
print(f"\nGünlük Üretim-Fiyat Korelasyonu: {daily_corr:.3f}")

# En değerli günlerde üretim performansı
top_price_days = daily_production.nlargest(30, 'PTF')
avg_production_high_price = top_price_days['URETIM'].mean()
avg_production_overall = daily_production['URETIM'].mean()
high_price_utilization = avg_production_high_price / avg_production_overall

print(f"En Yüksek Fiyatlı 30 Günde Ort. Üretim: {avg_production_high_price:.1f} MWh")
print(f"Genel Ortalama Günlük Üretim: {avg_production_overall:.1f} MWh")
print(f"Yüksek Fiyat Kullanım Oranı: {high_price_utilization:.2f}")

# Hidro strateji skoru
hydro_strategy_score = (price_utilization_index + daily_corr + high_price_utilization) / 3 * 100
print(f"\nRüzgar Strateji Skoru: {hydro_strategy_score:.1f}/100")
print(f"(Fiyat takibi, günlük korelasyon ve yüksek fiyat kullanımının ortalaması)")

=== EBER RES - RÜZGARLI Santral Fiyat Kullanım Endeksi ===
Fiyat Seviyelerine Göre Üretim Dağılımı:
            Avg_PTF  Total_Production  Hour_Count   Avg_DM
PTF                                                       
Çok_Düşük   1161.62          49025.24        1754  3370.08
Düşük       2076.86          38882.24        1756  2982.76
Orta        2374.00          38839.64        1758  2481.25
Yüksek      2629.03          39632.64        1753  1871.27
Çok_Yüksek  2939.62          37360.58        1757  1201.28

Fiyat Kullanım Endeksi: 0.975
(1.0'dan yüksek = yüksek fiyatlarda daha çok üretim)

Günlük Üretim-Fiyat Korelasyonu: -0.169
En Yüksek Fiyatlı 30 Günde Ort. Üretim: 372.5 MWh
Genel Ortalama Günlük Üretim: 556.8 MWh
Yüksek Fiyat Kullanım Oranı: 0.67

Rüzgar Strateji Skoru: 49.1/100
(Fiyat takibi, günlük korelasyon ve yüksek fiyat kullanımının ortalaması)


/var/folders/1t/15hmc_h12xnbygjnmf7kj43h0000gn/T/ipykernel_1820/2580323079.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  price_efficiency = df.groupby(price_levels).agg({
